OrbitCore/LinuxEventTracer.cpp

```diff
@@ -0,0 +1,193 @@
+        // we schedule the reading using round robbin like scheduling.
+        for (std::shared_ptr<LinuxPerfRingBuffer> ring_buffer : ring_buffers)
+        {
+            int i = 0;
+            // read everything that is new
+            while (ring_buffer->HasNewData() && i < ROUND_ROBIN_BATCH_SIZE) {
+                i++;
+                new_events = true;
+                perf_event_header header{};
+                ring_buffer->ReadHeader(&header);
+
+                // perf_event_header::type contains the type of record, e.g.,
+                //  PERF_RECORD_SAMPLE, PERF_RECORD_MMAP, etc., defined in enum
+                //  perf_event_type in perf_event.h.
+                switch (header.type)
+                {
+                case PERF_RECORD_SAMPLE:
+                    {
+                        auto sched_switch = ring_buffer->ConsumeRecord<sched_switch_record_t>(header);
+                        if ( GParams.m_SystemWideScheduling )
+                        {
+                            // record end of excetion
+                            if (sched_switch.raw_data.prev_pid != 0)
+                            {
+                                ++Capture::GNumContextSwitches;
+
+                                ContextSwitch CS ( ContextSwitch::Out );
+                                CS.m_ThreadId = sched_switch.raw_data.prev_pid;
+                                CS.m_Time = sched_switch.time;
+                                CS.m_ProcessorIndex = sched_switch.cpu;
+                                //TODO: Is this correct?
+                                CS.m_ProcessorNumber = sched_switch.cpu;
+                                GTimerManager->Add( CS );
+                            }
+
+                            // record start of excetion
+                            if (sched_switch.raw_data.next_pid != 0)
+                            {
+                                ++Capture::GNumContextSwitches;
+
+                                ContextSwitch CS ( ContextSwitch::In );
+                                CS.m_ThreadId = sched_switch.raw_data.next_pid;
+                                CS.m_Time = sched_switch.time;
+                                CS.m_ProcessorIndex = sched_switch.cpu;
+                                //TODO: Is this correct?
+                                CS.m_ProcessorNumber = sched_switch.cpu;
+                                GTimerManager->Add( CS );
+                            }
+                        }
+                        else
+                        {
+                            // we need to know the type of the perf_record_sample at compile time,
+                            //  i.e. for multiple perf_event_open calls, we need be able to
+                            //  distinguish the different ring buffers.
+                            event_buffer.Push(std::make_unique<LinuxSchedSwitchEvent>(sched_switch));
```

Comment:
It might be cheaper to copy the pointers in the priority_queue (when inserting a new element and moving the other elements around) instead of copying the complete objects. On the other side, the elements would be then all in the same range of memory (due to the underlying vector), which is more cache friendly.

We should do micro benchmarks here. For the moment, I'll stick with the pointers and add a comment here.